group.go

```diff
@@ -0,0 +1,463 @@
+		// be only one entry for this key.
+		//
+		// Consider the following serialized event ordering for two
+		// goroutines in which this callback gets called twice for hte
+		// same key:
+		// 1: Get("key")
+		// 2: Get("key")
+		// 1: lookupCache("key")
+		// 2: lookupCache("key")
+		// 1: load("key")
+		// 2: load("key")
+		// 1: loadGroup.Do("key", fn)
+		// 1: fn()
+		// 2: loadGroup.Do("key", fn)
+		// 2: fn()
+		if value, cacheHit := g.lookupCache(key); cacheHit {
+			g.Stats.CacheHits.Add(1)
+			return value, nil
+		}
+		g.Stats.LoadsDeduped.Add(1)
+		var value transport.ByteView
+		var err error
+		if peer, ok := g.instance.PickPeer(key); ok {
+
+			// metrics duration start
+			start := time.Now()
+
+			// get value from peers
+			value, err = g.getFromPeer(ctx, peer, key)
+
+			// metrics duration compute
+			duration := int64(time.Since(start)) / int64(time.Millisecond)
+
+			// metrics only store the slowest duration
+			if g.Stats.GetFromPeersLatencyLower.Get() < duration {
+				g.Stats.GetFromPeersLatencyLower.Store(duration)
+			}
+
+			if err == nil {
+				g.Stats.PeerLoads.Add(1)
+				return value, nil
+			}
+
+			if errors.Is(err, context.Canceled) {
+				return nil, err
+			}
+
+			if errors.Is(err, &transport.ErrNotFound{}) {
+				return nil, err
+			}
+
+			if errors.Is(err, &transport.ErrRemoteCall{}) {
+				return nil, err
+			}
+
+			if g.instance.opts.Logger != nil {
+				g.instance.opts.Logger.Error(
+					fmt.Sprintf("error retrieving key from peer '%s'", peer.PeerInfo().Address),
+					"category", "groupcache",
+					"err", err,
+					"key", key)
+			}
+
+			g.Stats.PeerErrors.Add(1)
+			if ctx != nil && ctx.Err() != nil {
+				// Return here without attempting to get locally
+				// since the context is no longer valid
+				return nil, err
+			}
+		}
+
+		value, err = g.getLocally(ctx, key, dest)
+		if err != nil {
+			g.Stats.LocalLoadErrs.Add(1)
+			return nil, err
+		}
+		g.Stats.LocalLoads.Add(1)
+		destPopulated = true // only one caller of load gets this return value
+		g.populateCache(key, value, &g.mainCache)
+		return value, nil
+	})
+	if err == nil {
+		value = viewi.(transport.ByteView)
+	}
+	return
+}
+
+func (g *group) getLocally(ctx context.Context, key string, dest transport.Sink) (transport.ByteView, error) {
+	err := g.getter.Get(ctx, key, dest)
+	if err != nil {
+		return transport.ByteView{}, err
+	}
+	return dest.View()
+}
+
+func (g *group) getFromPeer(ctx context.Context, peer peer.Client, key string) (transport.ByteView, error) {
+	req := &pb.GetRequest{
+		Group: &g.name,
+		Key:   &key,
+	}
+	res := &pb.GetResponse{}
+	err := peer.Get(ctx, req, res)
+	if err != nil {
+		return transport.ByteView{}, err
+	}
+
+	var expire time.Time
+	if res.Expire != nil && *res.Expire != 0 {
+		expire = time.Unix(*res.Expire/int64(time.Second), *res.Expire%int64(time.Second))
+		if time.Now().After(expire) {
+			return transport.ByteView{}, errors.New("peer returned expired value")
```

Comment:
Is the peer expected to eventually expire the value from their cache? Or would this Get() coax the peer to expire this element immediately?

I'm having trouble wrapping my head around why we'd return an error for an expired value vs. seamlessly expiring the element and treating it as a NotFound case, kind of how the LRU does an implicit removeElement() on Get(). Could you help me grok this one?

Reply:
Each peer should expire the item. This error is intended to catch a scenario where the peer's clock was behind the clients clock (or by the time the value transited the network it was expired) and the peer returned an expired item. As the code is currently written, `groupcache` will log the error and fall back to fetching the value locally.

Your comment here has caused me to re-evaluate inclusion of this error.  I think the **correct** thing to do is to accept the item even if it's expired. If we continue with the existing code, we could inadvertently cause a thundering herd scenario if all nodes in the cluster happen to hit this scenario because the owner's clock is behind the rest of the cluster.

Reply:
💡 makes sense! Thanks for the clarification!